core/src/bls_sigverify/bls_sigverifier.rs

```diff
@@ -0,0 +1,1565 @@
+                        pubkey: *solana_pubkey,
+                    });
+                }
+                ConsensusMessage::Certificate(cert) => {
+                    // Only need certs newer than root slot
+                    if cert.cert_type.slot() <= root_bank.slot() {
+                        self.stats.received_old.fetch_add(1, Ordering::Relaxed);
+                        packet.meta_mut().set_discard(true);
+                        continue;
+                    }
+
+                    if self
+                        .verified_certs
+                        .read()
+                        .unwrap()
+                        .contains(&cert.cert_type)
+                    {
+                        self.stats.received_verified.fetch_add(1, Ordering::Relaxed);
+                        packet.meta_mut().set_discard(true);
+                        continue;
+                    }
+
+                    certs_to_verify.push(cert);
+                }
+            }
+        }
+        preprocess_time.stop();
+        self.stats.preprocess_count.fetch_add(1, Ordering::Relaxed);
+        self.stats
+            .preprocess_elapsed_us
+            .fetch_add(preprocess_time.as_us(), Ordering::Relaxed);
+
+        let (votes_result, certs_result) = rayon::join(
+            || self.verify_and_send_votes(votes_to_verify),
+            || self.verify_and_send_certificates(certs_to_verify, &root_bank),
+        );
+
+        votes_result?;
+        certs_result?;
+
+        // Send to metrics service for metrics aggregation
+        if self
+            .consensus_metrics_sender
+            .send((Instant::now(), consensus_metrics_to_send))
+            .is_err()
+        {
+            warn!("could not send consensus metrics, receive side of channel is closed");
+        }
+
+        self.stats.maybe_report_stats();
+
+        Ok(())
+    }
+}
+
+impl BLSSigVerifier {
+    pub fn new(
+        sharable_banks: SharableBanks,
+        verified_votes_sender: VerifiedVoteSender,
+        message_sender: Sender<ConsensusMessage>,
+        consensus_metrics_sender: ConsensusMetricsEventSender,
+    ) -> Self {
+        Self {
+            sharable_banks,
+            verified_votes_sender,
+            message_sender,
+            stats: BLSSigVerifierStats::new(),
+            verified_certs: RwLock::new(HashSet::new()),
+            vote_payload_cache: RwLock::new(HashMap::new()),
+            consensus_metrics_sender,
+            last_checked_root_slot: 0,
+        }
+    }
+
+    fn verify_and_send_votes(
+        &self,
+        votes_to_verify: Vec<VoteToVerify>,
+    ) -> Result<(), BLSSigVerifyServiceError<ConsensusMessage>> {
+        let verified_votes = self.verify_votes(votes_to_verify);
+        self.stats
+            .total_valid_packets
+            .fetch_add(verified_votes.len() as u64, Ordering::Relaxed);
+
+        let mut verified_votes_by_pubkey: HashMap<Pubkey, Vec<Slot>> = HashMap::new();
```

Comment:
I think we're fine to make this a simple AHash.

I also wonder if there's any pre-allocation we should do here.

And finally, should we use a `Set` instead of a `Vec` here? We expect uniqueness, right?